### ISSUE TITLE
MI loss for pytorch

### ISSUE THREAD

**Issue by lemoshu** [voxelmorph/voxelmorph#259]
Hi,

Is there a PyTorch version for the NMI loss?

**Comment by adalca** [comment]
@lemoshu , 
I'm sorry, we don't have one right now. It would probably be easy to translate from tensorflow, especially the new implementation in https://github.com/adalca/neurite/blob/dev/neurite/tf/metrics.py, but unfortunately we haven't had time.

**Comment by junyuchen245** [comment]
Hi @lemoshu ,

I have translated @adalca 's MI code (https://github.com/adalca/neurite/blob/dev/neurite/tf/metrics.py) into Pytorch:
https://github.com/junyuchen245/ViT-V-Net_for_3D_Image_Registration_Pytorch/blob/b6a9c884ecebb0b0126be7c1aae607050ec54d70/ViT-V-Net/losses.py#L288

I did some quick tests, and the code seems to be working. I hope this helps.

**Comment by adalca** [comment]
@junyuchen245 This is cool -- would you mind also adding this to dev/neurite/torch/metrics.py as an open source PR? 

We could do it but it's helpful if the community contributes to the code and this way you'd get the sweet github credit. You should also add your username in the documentation. We may need to modify it upon testing.

**Comment by junyuchen245** [comment]
> @junyuchen245 This is cool -- would you mind also adding this to dev/neurite/torch/metrics.py as an open source PR?
> 
> We could do it but it's helpful if the community contributes to the code and this way you'd get the sweet github credit. You should also add your username in the documentation. We may need to modify it upon testing.

Thanks for offering. I just added it as a PR to dev/neurite/torch/metrics.py.

**Comment by adalca** [comment]
Great, thank you. We're going to take a look at it, like accept it as is and do some testing!

**Comment by junyuchen245** [comment]
Sure thing. I'm glad that I could help.

**Comment by timroelofs123** [comment]
Hi @junyuchen245 , I was wondering if you managed to successfully train a model with this loss function? I tried to use it but it doesn't work to me, the loss value doesn't change with the input it seems. Any help would be appreciated!

**Comment by junyuchen245** [comment]
Hi @timroelofs123 , 

I had a similar issue, but I double-checked the code, and it matched @adalca 's TensorFlow version.

Anyways, I have been using this simplified version, and it's working well for me. You can give this a try: https://github.com/junyuchen245/ViT-V-Net_for_3D_Image_Registration_Pytorch/blob/5ebef416a7bfefda80b4bd8f58e6850eccebca2e/ViT-V-Net/losses.py#L288.

Let me know how it goes.

Thanks,
Junyu

**Comment by timroelofs123** [comment]
That's what I thought too, it seemed to match the TF code and the algorithm as described by Guo (2019), so I'm not sure where the issue lies. However, your simplified implementation appears to work correctly, thanks a lot! I'll let you know if I run into any issues. 

By the way, @adalca I was wondering if there are plans to release the code for the local MI metric, for either TF or PyTorch?

**Comment by junyuchen245** [comment]
@timroelofs123 I'm glad to hear it works. I just uploaded the PyTorch version of the local MI loss (for non-overlapping local patches), which I implemented based on a previous TF version of VoxelMorph.

Here is the loss:
https://github.com/junyuchen245/ViT-V-Net_for_3D_Image_Registration_Pytorch/blob/19559170e915b44e181c83ffe9571533340e9320/ViT-V-Net/losses.py#L345

**Comment by timroelofs123** [comment]
Oh awesome, thanks! I'll try it out. @junyuchen245
